prov/tcp/src/tcpx_conn_mgr.c

```diff
@@ -474,171 +285,147 @@ static void handle_conn_handle(struct poll_fd_mgr *poll_mgr,
-{
-	struct poll_fd_info *poll_info;
+	cm_ctx = calloc(1, sizeof(*cm_ctx));
+	if (!cm_ctx)
+		goto err2;
 
-	tcpx_fabric->poll_mgr.run = 0;
-	fastlock_acquire(&tcpx_fabric->poll_mgr.lock);
-	fd_signal_set(&tcpx_fabric->poll_mgr.signal);
-	fastlock_release(&tcpx_fabric->poll_mgr.lock);
+	handle->conn_fd = sock;
+	handle->handle.fclass = FI_CLASS_CONNREQ;
+	handle->pep = pep;
+	cm_ctx->fid = &handle->handle;
+	cm_ctx->type = SERVER_RECV_CONNREQ;
 
-	if (tcpx_fabric->conn_mgr_thread &&
-	    pthread_join(tcpx_fabric->conn_mgr_thread, NULL)) {
-		FI_DBG(&tcpx_prov, FI_LOG_FABRIC,
-		       "cm thread failed to join\n");
-	}
+	ret = ofi_wait_fd_add(wait, sock, FI_EPOLL_IN,
+			      tcpx_eq_wait_try_func,
+			      NULL, (void *) cm_ctx);
+	if (ret)
+		goto err3;
 
-	while (!dlist_empty(&tcpx_fabric->poll_mgr.list)) {
-		poll_info = container_of(tcpx_fabric->poll_mgr.list.next,
-					 struct poll_fd_info, entry);
-		dlist_remove(&poll_info->entry);
-		assert(poll_info->flags & POLL_MGR_FREE);
-		free(poll_info);
-	}
+	return;
+err3:
+	free(cm_ctx);
+err2:
+	free(handle);
+err1:
+	ofi_close_socket(sock);
+}
 
-	fastlock_destroy(&tcpx_fabric->poll_mgr.lock);
-	fd_signal_free(&tcpx_fabric->poll_mgr.signal);
+static void process_cm_ctx(struct util_wait *wait,
+			      struct tcpx_cm_context *cm_ctx)
+{
+	switch (cm_ctx->type) {
+	case SERVER_SOCK_ACCEPT:
+		server_sock_accept(wait,cm_ctx);
+		break;
+	case CLIENT_SEND_CONNREQ:
+		client_send_connreq(wait, cm_ctx);
+		break;
+	case SERVER_RECV_CONNREQ:
+		server_recv_connreq(wait, cm_ctx);
+		break;
+	case SERVER_SEND_CM_ACCEPT:
+		server_send_cm_accept(wait, cm_ctx);
+		break;
+	case CLIENT_WAIT_FOR_CONNRESP:
+		client_wait_for_connresp(wait, cm_ctx);
+		break;
+	default:
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"should never end up here\n");
+	}
 }
 
-int tcpx_conn_mgr_init(struct tcpx_fabric *tcpx_fabric)
+void tcpx_conn_mgr_run(struct util_eq *eq)
 {
-	int ret;
+	struct util_wait_fd *wait_fd;
+	void *wait_contexts[MAX_EPOLL_EVENTS];
+	int num_fds = 0, i;
 
-	dlist_init(&tcpx_fabric->poll_mgr.list);
-	fastlock_init(&tcpx_fabric->poll_mgr.lock);
-	ret = fd_signal_init(&tcpx_fabric->poll_mgr.signal);
-	if (ret) {
-		FI_WARN(&tcpx_prov, FI_LOG_FABRIC,"signal init failed\n");
-		goto err;
-	}
+	assert(eq->wait != NULL);
 
-	tcpx_fabric->poll_mgr.run = 1;
-	ret = pthread_create(&tcpx_fabric->conn_mgr_thread, 0,
-			     tcpx_conn_mgr_thread, (void *) tcpx_fabric);
-	if (ret) {
-		FI_WARN(&tcpx_prov, FI_LOG_FABRIC,
-			"Failed creating tcpx connection manager thread");
+	wait_fd = container_of(eq->wait, struct util_wait_fd,
+			       util_wait);
 
-		goto err1;
-	}
-	return 0;
-err1:
-	fd_signal_free(&tcpx_fabric->poll_mgr.signal);
-err:
-	fastlock_destroy(&tcpx_fabric->poll_mgr.lock);
-	return ret;
+	num_fds = fi_epoll_wait(wait_fd->epoll_fd, wait_contexts,
+				MAX_EPOLL_EVENTS, 0);
+	if (num_fds < 0)
+		return;
+
+	for ( i = 0; i < num_fds; i++)
```

Comment:
@vkrishna While testing your new code, I discovered an issue in the function tcpx_conn_mgr_run().
After fi_epoll_wait() returns, you need to make sure that wakeup signals do not enter process_cm_ctx() and must be skipped. You can fix this with the following lines:

```
       for ( i = 0; i < num_fds; i++) {
               /* Skip wake up signals */
               if (&wait_fd->util_wait.wait_fid.fid == wait_contexts[i])
                       continue;

                process_cm_ctx(eq->wait,
                                 (struct tcpx_cm_context *)
                                 wait_contexts[i]);
       }
```

Reply:
Thanks @sydidelot for taking time checking the patch and fixing this issue. I will add it.